pages/XX-ai-methods/reproducible_prompts/Reproducible_Prompts.qmd

```diff
@@ -0,0 +1,299 @@
+Create a list from the text of {random_words} in random order
+```
+
+## Lets keep making things more reproducible
+We can Set a seed to try and get more reproducible results. This is an ollama
+parameter, so we provide it inside the `-m {}` arguments. Just add `"seed:42`
+after our `"base_url":"http://localhost:11434"` seperated by a comma. **Do not add any spaces**, as it can interfere with parsing the arguments.
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434","seed":42}
+Create a bullet list from the text of {random_words} in random order
+```
+
+## Let's keep tweaking the randomness
+- You can adjust to "Creativeness" of the responses with the `"temperature"` 
+- Default is 0.8, higher is more creative
+- You can set `"temperature":0` to give the least "creative" response which can help reproducibity.
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434","seed":42,"temperature":1.5}
+Create a list of 10 random words
+```
+
+```{python}
+%%ai ollama:llama3.2 -m {"base_url":"http://localhost:11434","seed":42,"temperature":0}
+Create a list of 10 random words
+```
+
+## What about passing the output to your python code?
+You may want to pass the output of a Juypter AI cell to your python code.
+The best way to do this is to call the contents of your output cell. First let's
+run an AI cell that outputs json. Python can then input the json as a python
+dictionary.
+
+```{python}
+%%ai ollama:llama3.2 -f json -m {"base_url":"http://localhost:11434","seed":42,"temperature":0}
+Create a list of 10 random words
+```
+
+To capture the output of this cell You can call the `Out[##]` function, substituting `##` for the cell number. This can be tricky because the cell
+number changes based on the order of cells run or if you rerun a cell. The `_`
+operator will return the output of the last run cell. As long as you run two
+cells one after the other, it should always successfully retrieve the output.
+The output will likely be a `IPython.core.display.JSON` object which has the
+contents of the json saved as a python dictionary in the .data property.
+
+Lets call the retrieve the output.data `_.data` and save it to a object `rando`.
+
+```{python}
+rando = _.data
+type(rando)
+```
+
+```{python}
+type(Out[13])
+```
+
+We can select an element of the dictionary using a key.
+
+```{python}
+rando["word2"]
+```
+
+Using a key can be tricky, as we can't be %100 certain what keys our LLM will
+choose. It can be useful instead to index by position. To do this we need to
+get the list of keys using `.keys()`, coerce it into a `list()` and then select
+and element by index `[#]`. This will give use the key at position `#`.
+
+```{python}
+rando[list(rando.keys())[3]]
+```
+
+# Langchain
+## What about embeding LLM prompts directly into Python Code?
+Thats what langchain is for! Lets load a few packages; `pandas`, and `ollama`,
+`pydantic`
+
+```{python}
```

Comment:
We usually do our package imports up top. It's nice to acknowledge them when they're needed, and it flows better. But, a lot of times people want to see all the imports together so that they know what they need to install. I realize you do say that as well, but just for consistency I think we should stick with early on (maybe just a quirk from my days of writing Python scripts...).